### ISSUE TITLE
Remove EQ_TYPE judgment now?

### ISSUE THREAD

**Issue by jonsterling** [RedPRL/sml-redprl#407]
@favonia Now that we have a top level as a façon de parler (not ontologically real), is it possible to remove the type judgments?

**Comment by favonia** [comment]
Not until we change how levels are provided/synthesized, because `A in (U omega)` is never provable even if `a type at omega` is.

**Comment by jonsterling** [comment]
@favonia OK!

**Comment by favonia** [comment]
I have had enough about `(U omega)` with all the code duplication and #444 and others. Could we demand users to always specify a good upper bound on levels (from the outside) and restrict the proof theory to the fragment which disallow (directly) proving that `($ f m)` is of a type in a universe smaller than where the type of `f` belong? Right now, it is either all in level `#lvl` or `0` which seems compatible with this restriction.

PS: It still seems possible to prove that indirectly, by proving the well-typedness of ($ f m) in a larger universe, then the well-kindedness of that type in a smaller universe.

**Comment by jonsterling** [comment]
I'd really like to get rid of `EQ_TYPE` too. Can you try to explain your proposal a bit more?

Are you proposing to remove `omega`?

**Comment by favonia** [comment]
Yes. The proposed plan consists of...
1. The user has to provide a universal upper bound that works in a top-down manner. Whenever we have to synthesize some level, we use the provided upper bound.
2. Goodbye to the level `omega`.

**Comment by jonsterling** [comment]
@favonia I am worried about this upper-bound idea... Where would the user specify this?

**Comment by favonia** [comment]
When the user specify the theorem, it needs to be something like
```
  (->
   [ty/a ty/b : (U #l kan)]
   [p : (path [_] (U #l kan) ty/a ty/b)]
   ty/a
   ty/b) at (++ #l)
```
Note the `at (++ #l)` at the end.

**Comment by jonsterling** [comment]
@favonia I see... It seems really arbitrary, and I'm not sure I'm in favor of this idea.

I wonder, why not revamp the rules so that `omega` works properly everywhere, and then get rid of `EQ_TYPE`? Does that seem impossible?

It is very nice to be able to say that something is a type, without needing to know its level precisely. It seemed like, with some changes to the rules, we could use `(U omega)` as a façon de parler for this purpose, and things would just work seamlessly. But maybe I missed something crucial...

**Comment by favonia** [comment]
To show two elements are equal `(U omega)`, under the current semantics, you need to say `(U omega)` is well-kinded, which is not. One idea might be that, "if there is a way to replace each `(U omega)` by `(U l)` for some level `l` then it's fine" but then `(U omega) : (U omega)` is formally justified and probably something goes wrong in this direction...? @cangiuli

**Comment by favonia** [comment]
Hmm. Unless we implement some level-checking algorithm somewhere, it seems the replacements could be inconsistent when combining different parts of an expression (of a realizer).

**Comment by favonia** [comment]
(Integrating the deleted comments) @jonsterling Under the current semantics, `(U omega)` needs to be a type for `blah : (U omega)` to make sense, and then implicitly you have `(U omega) : (U omega)`. When an expression or a proof is full of `(U omega)`, we probably need a separate algorithm to check feasibility of consistent level assignments. Do you want that, then? The down side is that the algorithm makes the refinement deviate from the actual proof, and in the same direction we can very well support recursive expressions and other things.

**Comment by jonsterling** [comment]
@favonia Ahh you're right, my silly idea doesn't work 😄  Without a transfinite hierarchy, I don't think that what I was thinking can really work, and we probably don't want to do that.

I think that we should think a bit more about our options then.

**Comment by favonia** [comment]
@jonsterling Do you mean the level `(++ omega)`? That might work, though we probably still have to add `at foo` for custom operators used in other theorems.

It seems to me that it is an issue of _when_ the levels are assigned and how they are synthesized. From this point of view...
1. Assigned immediately from the outside and blindly synthesized from the outer constraint: one-time effort for each theorem and no code duplication.
2. Assigned when needed but remaining unspecified as long as possible: duplicate code for the fragment with unspecified levels (current).
3. Synthesized by some algorithm and assigned when overwriting needed: Nuprl.
4. Synthesized to omega for every top-level and assigned when overwriting is needed: similar to 1 except that we need an (++ omega) level and the overwriting is probably frequently needed.

We could massage the syntax so that it could be
```
Thm X (...) at 2 : [
  ...
] by [
  ...
].
```

**Comment by jonsterling** [comment]
@favonia Thank you for laying out the design space so clearly! Of the four alternatives that you have suggested, which do you think is the best or least frustrating? I assume you prefer (1)?

**Comment by favonia** [comment]
My hatred toward code duplication makes (1) or its variants my favorite(s).

**Comment by jonsterling** [comment]
@favonia OK, I'm closer to being convinced of it 😄

**Comment by favonia** [comment]
How should we proceed in this discussion? I wish we could reach some agreement so that various issues would have definite answers (wontfix or todo, for example). @cangiuli

**Comment by jonsterling** [comment]
@favonia I think we should try the idea you prefer.

**Comment by cangiuli** [comment]
I don't really understand what's going on here, because I'm not sure what's currently implemented.

**Comment by jonsterling** [comment]
@favonia I was thinking about level `omega` again, and I'm not sure I understand the the reasoning by which we decided it didn't suffice to resolve these problems.

You say that if we prove `A \in U\omega` (or `U\omega true`), then this entails that `U\omega` is a type, and that this means `U\omega \in U\omega`, which of course we can't have. But I don't think this follows, or at least, I think that we can design our proof theory so that it doesn't follow. Let me try to give an idea:

1. In the PRL, when we need to prove that `A` is a type, we approximate this by proving `A\in U\omega`.

2. In the PRL, from a hypothesis `A true` we could choose *not* to have a rule that lets you conclude `A \in U\omega`.

It's not completely clear to me whether the above scheme would be usable, but I think it at least can be made to be consistent. @favonia What do you think?

**Comment by favonia** [comment]
@jonsterling This means we don't want the user to prove `A type` from `x : A`. I believe this effectively means `x : #l` does not require `#l` to be a type. Another hypothesis, such as `#l : (U 0)`, is needed.

**Comment by jonsterling** [comment]
@favonia Yes; in this design, `A type` would be a notation for `A \in U\omega`, and we would not be able to conclude this from `x : A`. I don't know if that breaks everything usability-wise...

**Comment by favonia** [comment]
In more details: When `A` is a variable, it is probably bound somewhere, and hopefully the binder tells us the level. When `A` is an expression, the rules should decompose it to smaller pieces. The only thing that would be affected, as far as I can see, is meta-variables. Now membership needs to be explicitly passed.

**Comment by jonsterling** [comment]
@favonia You are probably right, but I didn't quite understand your remark. Could you give some example?

**Comment by favonia** [comment]
```
Thm Foo (#l) : [
  (-> [y : #l] (mem (path [_] #l) (abs [_] y)))
]
```
is probably not provable.

**Comment by favonia** [comment]
Hmm you are right. I need a better example, but you probably see the (potential) issue?

**Comment by jonsterling** [comment]
I believe there is a potential issue, but I can't quite put my finger on it.

**Comment by jonsterling** [comment]
I don't want to de-rail this discussion, but let me add the following:

This is really a case where unification-based existential variables would be helpful; the paradigmatic case for these level difficulties is the rule that proves `(lam [x] M) \in (-> A B)`; here we need to prove that `A` is a type, but we really don't care at what level (so we either need to pick a level, or we need to have a proof theory in which we don't need to specify the level).

One (not totally desirable) way to deal with it would be to use dependent subgoals:

```
H >> (lam [x] M) \in (-> A B)      [by fun/eq/lam]
    #0. H, x : A >> M \in B
    #1. H >> TERM(lvl)
    #2. H >> A\in (U #1)
```

But the "mode" here isn't quite desirable; we will only really figure out what `#1` is by trying to prove `#2`. Using tactics, we can kind of simulate that behavior, but it's not ideal.

But maybe something along these lines is an approach that we should use. This seems related to the issue about "temporary failure" that you mentioned, except where in that idea we want to signal a "temporary failure", in this example we want to signal a "contingent success". These are probably both instances of the same thing.

**Comment by favonia** [comment]
How about changing the semantics of `x : a` to "if `a` is a type, then `x` is a member of `x`"? @jonsterling @cangiuli

**Comment by favonia** [comment]
Hmm this seems very chaotic especially when we want to internalize equality judgments. What's really assumed in the following sequent?
```
_ : (= (U omega) a b) >> J
```

**Comment by jonsterling** [comment]
Yes, I'm worried about what it means to internalize such a thing.

**Comment by jonsterling** [comment]
I kind of want to avoid that question for now.

**Comment by favonia** [comment]
Alright. What is the current plan, then? #466 is not totally hopeless but (1) somehow it is very slow and (2) we need a level synthesizer so that most cases are automatic. I suppose we should grab other fixes inside #466, but what about the issues in #444?

**Comment by jonsterling** [comment]
@favonia OK; I think that we should try and extract whatever progressive aspects of #466 can be used, but wait and think some more before we try to resolve the main issues that #466 hoped to deal with.

The speed problem (which is in some cases more than 500% increase) is probably too much to be able to accept it as-is.

**Comment by favonia** [comment]
Alright, so, since #466 is not satisfactory, here is my proposal:

1. Keep `EQ_TYPE`, but remove all level annotations as in #494. `EQ_TYPE` will have the implicit omega level. After this, we can remove `LOMEGA`.
2. Implement the rules in a way that can handle both `EQ_TYPE` and the internalized equality in `(U l)`.
3. The `SUB_TYPE` is probably needed to handle level/kind mismatches elegantly. Perhaps we need `SUB_KIND` to match `EQ_TYPE` as well.

The downside is code duplication, but maybe it will not be so bad after making each rule work for both cases.

**Comment by favonia** [comment]
This is also partially addressed by #502.

**Comment by favonia** [comment]
Now all the preparation is done, and the code duplication is managed by functions in `View` and `Wrapper`.  The only exception is that somehow `SUB_UNIVERSE` was not renamed to `SUB_KIND` due to my stupidity. @jonsterling Do you feel strongly about the renaming? In any case, the remaining question is whether we want `EQ_TYPE` to hold type equality at the omega level.

**Comment by jonsterling** [comment]
@favonia I would like to rename it to `SUB_KIND` if you're cool with it.  I'm not sure about `EQ_TYPE`, what is your opinion?
